### ISSUE TITLE
Request for `uv.lock` to support different index urls across different developer machines and CI environments

### ISSUE THREAD

**Issue by humanzz** [astral-sh/uv#6349]
Hello team,

Many thanks for the awesome work you're doing, and the countless amounts of time you've saved everyone using `uv` (and `ruff`).

I've seen the recent blog post about 0.3.0 and the accompanying documentation update, and from playing a bit to test commands like `uv lock` and `uv sync`, I actually have a question about `uv.lock`.

I've described the setup I use in https://github.com/astral-sh/uv/issues/1710 (a proxy, and using `invoke`, etc.). That setup means that on different team member's machine's, each is likely to have a different `INDEX_URL` and even the CI/CD system we use would have a different one.

I've generated the `uv.lock` file using `UV_INDEX_URL=... uv lock`, and the main thing I noticed is that the index url is included in the lock file e.g.

```
[[package]]
name = "aws-embedded-metrics"
version = "3.2.0"
source = { registry = "http://<omitted>/pypi/simple" }
dependencies = [
    { name = "aiohttp" },
]
sdist = { url = "http://<omitted>/pypi/simple/aws-embedded-metrics/3.2.0/aws-embedded-metrics-3.2.0.tar.gz", hash = "sha256:f235f87ab25ff328f6f3afca1c6b3218e81eea6e96e6aee012d368bb813fae7b" }
wheels = [
    { url = "http://<omitted>/pypi/simple/aws-embedded-metrics/3.2.0/aws_embedded_metrics-3.2.0-py3-none-any.whl", hash = "sha256:887b76d24914efa5fc42a7b77983e77fc670633e6e1195aac7653c425fee7399" },
]
```

I have not tried using that file across our different machines yet, but I suspect it's going to cause problems so I wanted to confirm with you folks
- What would happen if the index in the lock file is not available, and `UV_INDEX_URL`?
- Is there a way to not include the full index url in the lock file?

**Comment by humanzz** [comment]
I've tested it further and confirmed that `uv sync` would fail on the different machines given the `UV_INDEX_URL=` would be pointing to a different index url.

With that that, I think this is a request to enable not writing the index url into the lock files via `uv lock`, and to enable `uv sync` - and other commands to use the lock file that does not have index urls as part of the lockfile

**Comment by mgab** [comment]
I discovered `uv` recently and I'm pretty excited about it! So even if we're discussing issues, thanks! 🫶

Now, coming to the topic, I think this becomes important given the ability to setup `index-url` and `extra-index-url` as user-level configuration, and potentially as system-level configuration (#6742).

Right now if you have a user-level config pointing at a package repository that acts as proxy to PyPI (that might add private pacakges on top) you will record that url in any lock file you generate. So if you want to work on some public project where `uv` is used as pacakge manager, either

1. the project -that you might not control- specifically defines `index-url` and `extra-index-url` to point at `PyPI`
2. you disable the user-level config 
3. or you will generate lock files that people outside of your org will not be able to use

Option 2 might not seem that bad, but then that defies the purpose of being able to define user-level or system-level configs. Plus, it might not be the ideal setup ideal to take advantage of `uvx` and `uv tool` using private packages that are only available in the private repository.

**Comment by humanzz** [comment]
Just adding a couple more comments
- I did see on some other issue/pr, a mention that thanks to including those full paths, `uv` wouldn't be doing any further resolutions, and it just goes ahead to install things - which would save some time which sounds great
- In the case of the setup we're using, I noticed something which might be useful to call out
  - when the index, is public pypi index i.e. `https://pypi.org/simple`, the package artifacts have `url`s with a prefix looking like `https://files.pythonhosted.org/packages/c1/08/1bb1f7392cd9c3a65c4571a7f7286056ec52b3cd3d432485ca2b9907e0f9/`. For example, for `aws-embedded-metrics`, it's `https://files.pythonhosted.org/packages/c1/08/1bb1f7392cd9c3a65c4571a7f7286056ec52b3cd3d432485ca2b9907e0f9/aws-embedded-metrics-3.2.0.tar.gz`
  - in my case, this custom index, which uses a local proxy on dev/ci machines i.e. `http://<omitted>/pypi/simple`, the package srtifacts `url`s follow a more deterministic url pattern e.g. `http://<omitted>/pypi/simple/<package name>/<package version>/`. The same `aws-embedded-metrics` would be `http://<omitted>/pypi/simple/aws-embedded-metrics/3.2.0/aws-embedded-metrics-3.2.0.tar.gz`.

This determinism, can actually lead to a simpler handling of this, where I tested with
1. Generate the `uv.lock` with setting a `UV_INDEX_URL` to my custom index
1. Used it in other machines, by running another small script, prior to `uv sync`, that modifies `uv.lock` to replace the index in `http://<omitted>/pypi/simple` with the content of `UV_INDEX_URL`

All of this, makes me wonder if there's some solution to cases like mine, along the lines of
- Executing commands like `uv lock`, to have a flag to indicate usage of placeholder index urls, where it would have a place holder for `UV_INDEX_URL` or the extra indices
- Executing commands like `uv sync` can take a similar flag, where it'd use the values passed through `UV_INDEX_URL` to resolve the placeholder values in `uv.lock`

**Comment by smheidrich** [comment]
What follows is a +1 comment you should all ignore, but I'm writing it to include some keywords in this issue so other people who run into this can find it more easily than I just did (took me a while):

-----

Having the same issue because I fetch all my PyPI packages through a local [devpi](https://github.com/devpi/devpi)<sup>keyword&nbsp;1</sup> installation, so all `source.registry` and `sdist.url` URLs in `uv.lock` say `https://localhost:3141/...`<sup>keyword&nbsp;2</sup> and, needless to say, this results in `Connection refused`<sup>keyword&nbsp;3</sup> errors when trying to install the package elsewhere (e.g. CI systems such as [GitHub Actions](https://github.com/smheidrich/demo-sphinx-paramlinks-sphinx-autoapi-compat/actions/runs/10967757294/job/30458100347)<sup>keyword&nbsp;4</sup> or GitLab CI/CD<sup>keyword&nbsp;5</sup>).

**Comment by humanzz** [comment]
The more comments this gets, the more I think there needs to be an option for the lock file to not include any index information.

This way, all those use cases that use local proxies, different on developer machines from those in CI environments, and the likelihood that those different proxies having different conventions for their asset (wheel/sdist) URLs, that a general simple solution to exclude those index/asset URLs would be very useful and enable them to adopt `uv.lock`

**Comment by tapetersen** [comment]
We could really also use this to be able to have lock-files without disclosing internal urls.
If not in the sync file then an option to the `export` subcommand to export a new uv.lock but with this or other things omitted would also be a good alternative

**Comment by tapetersen** [comment]
@charliermarsh That may very well work for our requirements but it looks like the uv lockfile is richer (all the extras are declared). Otherwise it would just be a convenience to not have someone outside (that don't have access to internal dns-names) being able to use the lock-file without having to regenerate it from the requirements.txt (pardon me if I've missed a quicker way to handle that).

Basically my thought was that the hashes should be enough to guarantee that it's indeed the same packages that are being installed even if the urll has to be found again.

That said if it doesn't feel like the right way to do it and the recommended way is to manage that with requirements.txt exports we'll try that and raise a more specific issue or discussion if a showstopper or question regarding that comes up.

(feel free to edit/move/hide post if it derails original thread)

**Comment by humanzz** [comment]
@charliermarsh 

At the moment, I'm actually using `uv pip compile pyproject.toml...` and `uv pip install -r ...` to simulate having a lockfile.

The main thing I don't like about it, is keeping the lockfile up to date with `pyproject.toml` dependencies is all manual. I didn't want to always call `uv pip compile`, then follow immediately by `uv pip install` as that would (1) be slower (2) force updating resolved dependencies that are declared via version ranges.

So, I'd say that's not on par with the experience of something like `package.json` and `package-lock.json` for node, and I was hoping that `uv.lock` and the commands that maintain/use it are giving a better experience.

The other thing I like about `uv.lock` is `uv.tool.dev-dependencies`. Without that, I end up declaring those dependencies as optional dependencies which is really not right, but ended up doing it this way, so I can declare all my dependencies in `pyproject.toml`, and then generate the current `requirements.txt` "lockfile".

Omitting the URLs is one potential solution - me thinking out loud.

In the case of my proxy, given that the `sdist`/`wheel` URLs have a specific pattern with the index url being a prefix, I was able to run some code to manipulate the lock file before running `uv sync --all-extras --frozen` to install dependencies

```python
@task
def update_lock_index_url(ctx):
    """
    Workaround for `uv lock` writing the index url to the lock file.
    Update `source` and `url` fields with values based on `UV_INDEX_URL` environment variable.
    """
    with open("uv.lock", "r+") as file:
        contents = file.read()
        pattern = r"http://127.0.0.1:\d+/[^/]+/[^/]+/[^/]+/pypi/simple"
        replacement = os.environ.get("UV_INDEX_URL", "NO_INDEX_OVERRIDE_SET")
        new_contents = re.sub(pattern, replacement, contents)
        file.seek(0)
        file.truncate()
        file.write(new_contents)
```

Downside for this, is that any dev on my team will end up having modifying `uv.lock`.
2 thoughts I had about this
1. `uv` having this concept of a hook - some python code to execute to modify file content - before installing dependencies
2. Ability to specify path to a lockfile, whereby I can run the hook myself, to generate a modified version of the lockfile, make sure it's in .gitignore, or in some tmp directory, and use `uv sync` while pointing to my modified version

I thought the simplest of all of the above would be this whole omission of URLs from the file tbh.
Moreover, generally when I think about what's uv story for supporting specifying index/extra index, and how that interacts with commands like `uv sync`, I think the strong assumption there is that the index URLs are not changing, which at least in my case is not true - whether across dev machines, or in the CI environment.

**Comment by mgab** [comment]
> I'm not sure that a lockfile without URLs is a good idea. One of the goals of the lockfile is to be hermetic: you can install from just the lockfile without making any queries or relying on any external sources.
> 
> If you're going to omit source information like that, why not just export to requirements.txt?

As far as I can see, poetry records the file name and the hash, which should be enough to ensure that you are getting exactly the same files without relying on external sources.

On the other hand, by including the URLs you are coupling the lock file with the specific path of the package repository that you are using. And while this might remove the need of interacting with the package index, it also seems that adding a proxy package repository on the working environment should be transparent to the dependency specification of my project.

**Comment by charliermarsh** [comment]
I consider it fairly critical that the lockfile includes the URLs directly and I suspect that the lockfile standard (should it be accepted) will do the same. We do have a plan to support these use-cases though. After #7481, we'll add something like:

```toml
[[tool.uv.index]]
name = "private"
url = "https://private.org/simple"
proxy = "http://<omitted>/pypi/simple"
```

So you can set the index URL (which will appear in the lockfile) and then the proxy separately. You'll also be able to set the proxy as an environment variable, like `UV_PRIVATE_PROXY_URL` or similar.

**Comment by humanzz** [comment]
Hey @charliermarsh,

Just to make sure I understand this, and try to assess whether it'd work.
- The `url` in our case is not needed, though I dunno whether the intention is for it to be a valid index, or any url so maybe you can clarify that.
- I guess the `url` is what would be saved in index? and the `proxy` is kinda what replaces it when doing the actual resolving/downloads?
- for the package urls e.g. `sdist`, and `wheels`, how would that work with the `proxy`? In my earlier messages, I did see `https://pypi.org/simple` has `sdist` and `wheels` that don't have the index as part of their urls e.g. `https://files.pythonhosted.org/packages/c1/08/1bb1f7392cd9c3a65c4571a7f7286056ec52b3cd3d432485ca2b9907e0f9/aws-embedded-metrics-3.2.0.tar.gz` but in my case, it does have the index in the url e.g. `http://<omitted>/pypi/simple/aws-embedded-metrics/3.2.0/aws-embedded-metrics-3.2.0.tar.gz`

**Comment by charliermarsh** [comment]
1. Ideally it is a valid index and indicative of the actual locations of the files but that wouldn't be strictly required.
2. Correct.
3. Both the file URLs and the registry URLs would be "replaced" correctly as long as they're on the same domain. So this wouldn't work for PyPI but would work for your case.

**Comment by humanzz** [comment]
Great... then I look forwards to that :)

**Comment by mgab** [comment]
> I consider it fairly critical that the lockfile includes the URLs directly and I suspect that the lockfile standard (should it be accepted) will do the same.

Sorry! Actually, reading about the proposed standard I found [this comment](https://discuss.python.org/t/pep-751-lock-files-again/59173/57) that properly describes what was my concern, and then the answer to it. The point that I was missing is in fact addressed in the [PEP draft](https://peps.python.org/pep-0751/#packages-files-url):

>  - Installers MUST NOT assume the URL will always work, but installers MAY use the URL if it happens to work.

So yeah, I understood that URLs would be assumed to work, and that didn't seem a great idea. I see that's not the case, so nothing to object!

What got me confused is that apparently `uv sync` does assume the URLs to work unless you explicitly configure a value for either `index-url` or `extra-index-url` somewhere. I reported it in #8076 and #8074, hope it's useful. In any case, that's just about keeping improving `uv` and not a design disagreement.

Thanks for the answers!

**Comment by humanzz** [comment]
I saw that the revamped index support got released in 0.4.23.

Is there any separate issues to track the proxy/env variables support, or is this issue itself that?

**Comment by charliermarsh** [comment]
I think it’s ok to track it here.

**Comment by rd-danny-fleer** [comment]
May I ask what's the current status of the issue? What aspects remain unclear?

Currently I use a workaround (Run `uv lock` once with the alternative index URL and use the updated `uv.lock` file in the whole pipeline) to make CICD work with an alternative index. However, with this approach, you must be careful not to update any dependencies.

**Comment by charliermarsh** [comment]
I don't know if anything is unclear, it's more that it hasn't been prioritized.

**Comment by dariocurr** [comment]
Any updates on this? This is critical for us to ship `uv` in our pipelines

**Comment by jtfmumm** [comment]
I've [created a draft PR](https://github.com/astral-sh/uv/pull/11782) for this issue. The current design adds a new command line arg (`--index-proxy-url <index-name>=<proxy-url>`) and env var (`UV_INDEX_PROXY_URL=<index-name>=<proxy-url>`). 

You define a canonical URL for an index by defining the index the normal way in `pyproject.toml`:

```
 [[tool.uv.index]]
name = "private"
url = "https://private.org/simple"
```

When passing in a proxy URL for an index, that URL will be used for making calls to the registry. However, only the canonical URL will be written to the lockfile. If the canonical URL is also valid, then you can run commands like `uv sync` without a proxy URL. If it's not valid, you will need to pass in the proxy URL.

This is supported for `uv run`, `uv add`, `uv sync`, and `uv lock`.

I'd be curious to hear any feedback about whether this addresses your concerns and whether the API makes sense. I'd also be curious if anyone would be interested in trying out my branch with their projects.

**Comment by apollo13** [comment]
Hi @jtfmumm, thank you for your work on this! Since you asked for feedback… ;)

> The current design adds a new command line arg (--index-proxy-url <index-name>=<proxy-url>) and env var (UV_INDEX_PROXY_URL=<index-name>=<proxy-url>).

This works only for a single index if I understand the code correctly? I guess this will work for the majority of cases and at least `--index-proxy-url` could be specified multiple times. Not sure how that would work for `UV_INDEX_PROXY_URL` though.

> One tricky aspect (https://github.com/astral-sh/uv/issues/6349#issuecomment-2393167729) is that this will only work if your index uses its URL as the base URL for the source distributions and wheels. So as Charlie https://github.com/astral-sh/uv/issues/6349#issuecomment-2393409286, this doesn't work for PyPI, for example.

That is the biggest downside for me. We usually try to mirror PyPI to save bandwidth (for PyPI). I wonder if supporting something like PDM has would make sense: https://pdm-project.org/latest/usage/lockfile/#static-urls -- PDM has an option to only store the filename in the lockfile which probably means more requests to the index at some point (?).

**Comment by jtfmumm** [comment]
Hi @apollo13, thanks for the feedback! 

Currently you can specify `--index-proxy-url` multiple times. We're going to change the env var to be of the form `UV_INDEX_<index-name>_PROXY_URL. That is how we currently handle username and password for an index.

> PDM has an option to only store the filename in the lockfile

Charlie raises the concern [above](https://github.com/astral-sh/uv/issues/6349#issuecomment-2392197432) that URLs may be required once the lockfile standards are accepted, so I'm not sure if this option is open to us.

**Comment by apollo13** [comment]
> Charlie raises the concern [above](https://github.com/astral-sh/uv/issues/6349#issuecomment-2392197432) that URLs may be required once the lockfile standards are accepted, so I'm not sure if this option is open to us.

Understood, I wonder if we can come up with other creative ideas. Because using a proxy for PyPI in CI is the *one* situation where I see massive bandwidth savings for PyPI.

**Comment by charliermarsh** [comment]
> Understood, I wonder if we can come up with other creative ideas. Because using a proxy for PyPI in CI is the one situation where I see massive bandwidth savings for PyPI.

When you proxy PyPI in this way, are you also proxying the returned file URLs? Or just the Simple index API? Either way, I think this should still be fine with the proposed API, since... if you're just using the file URLs returned by PyPI, they don't need authentication; and if you're proxying those, then in theory you can use a domain that's the same as the top-level Simple API domain.

**Comment by apollo13** [comment]
The returned URLs are rewritten. Ie it is not a simple HTTP proxy but a repository manager like sonatype nexus, resulting in this download link for a package:
```
https://my_host/repository/pypi-proxy/packages/packaging/24.2/packaging-24.2-py3-none-any.whl
```
which makes the URL pretty much different from PyPI which has it at:
```
https://files.pythonhosted.org/packages/88/ef/eb23f262cca3c0c4eb7ab1933c3b1f03d021f2c48f54763065b6f0e321be/packaging-24.2-py3-none-any.whl
```

**Comment by apollo13** [comment]
I guess so, ie the simple index URL for the package `packaging` would be:
```
https://my_host/repository/pypi-proxy/simple/packaging
```
(my_host being the same as above). But I still don't understand why you think this would work. If I use the proxy to create the lockfile it would write package urls ala `.../packaging/24.2/packaging-24.2-py3-none-any.whl` into the lockfile and I don't think replacing that to the canonical URL would work.

**Comment by charliermarsh** [comment]
Ah I see. So you want to use a proxy for your own development, but put public URLs in the lockfile itself. That seems much more challenging, since we'd need to query PyPI for every package to get those URLs.

**Comment by humanzz** [comment]
Hi @jtfmumm!
I have no clue how I missed this, but many thanks for starting to work on this.

In my case, the `UV_INDEX_URL` is set to the proxy url, that `uv` uses. The value of that env variable will different between individual developers machines, and also on CI/CD environment. This means I cannot use a named index and a corresponding proxy index via `UV_INDEX_<index-name>_PROXY_URL`. I need to be able to use both `UV_INDEX_URL` and `UV_INDEX_PROXY_URL` so that the actual resolution happens from `UV_INDEX_URL` but the lock file gets written using `UV_INDEX_PROXY_URL`.

**Comment by ngnpope** [comment]
I've bumped into this at work today, but with a slightly different scenario. We have a private repository and proxy/mirror for PyPI. These are provided for local development via a VPN _(on an "internal" URL)_ and via mTLS for CI _(on an "external" URL)_. When accessing over the VPN we don't need to provide credentials for the repositories, but we do when accessing in CI over the external URL.

A couple of options came to my mind initially regarding a solution for this:

1. Provide an environment variable for the hostname

    - This has the disadvantage that the environment variable would always have to be set, unless it were possible to provide a default, e.g. `${INDEX_HOST:-repo.example.net}`. This sort of shell-like expansion is supported, for example, in Dockerfile.
    - And because that would need to be copied for every URL in the lock file it would quickly get messy...
    - Would depend on #5734

2. Provide something like git's [`url.<base>.insteadOf`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-urlbaseinsteadOf)

    - This would make it possible to drop a configuration file in CI providing a mapping of the base URL
    - It could also address the concerns around being able to provide a mapping for other URLs for source distributions and wheels
    - The downside is that it would be hard/messy to be able to do this via environment variables, so may only be limited to configuration files

I then discussed this with some colleagues and was helpfully pointed to this issue. Having looked through, it seems that Charlie's [original proposal](https://github.com/astral-sh/uv/issues/6349#issuecomment-2392197432) would have solved my issue, e.g. something like:

```toml
[[tool.uv.index]]
name = "mirror"
url = "https://repo.corp.net/mirror/simple"
default = true

[[tool.uv.index]]
name = "private"
url = "https://repo.corp.net/private/simple"
```

And then being able to override in CI with:

```sh
UV_MIRROR_PROXY_URL="https://repo.ext.corp.net/mirror/simple"
UV_PRIVATE_PROXY_URL="https://repo.ext.corp.net/private/simple"
```

As for my other suggestion number two above, that could look something like:

```toml
[[tool.uv.index.remap]]
"https://repo.corp.net/" = "https://repo.ext.corp.net/"

# We can even remap URLs for hosted files:
"https://files.pythonhosted.org/" = "https://some.other.mirror/"
```

(That's a rather rough example, but hopefully gets the point across.)

**Comment by ArcticLampyrid** [comment]
**FYI, here’s a summary of how another ecosystem—Poetry—handles this:**

1. It does *not* include the concrete sdist/wheel URLs in the lockfile; it only uses the SHA‑256 checksum.
2. It records the package *source*, unless it’s using the official PyPI source. If no source is recorded, it always defaults to PyPI’s official source.
3. There’s a plugin called [poetry-plugin-pypi-mirror](https://github.com/arcesium/poetry-plugin-pypi-mirror) that can replace the default source with a mirror at the project or user level (this replacement does *not* end up in the lockfile).

---

**IMO**, sdist/wheel URLs are *not* permanent—they’re internal implementation details of PyPI. You should always fetch the latest URL from the index file. For reproducibility and integrity checks, using the SHA‑256 checksum alone is sufficient.

In ecosystems like NPM, URLs are recorded because they’re fully canonicalized:

```
https://registry.npmjs.org/typescript/-/typescript-5.4.3.tgz
```

But PyPI’s URLs are *not* like that; they look more like opaque object storage identifiers:

```
https://files.pythonhosted.org/packages/c1/08/1bb1f7392cd9c3a65c4571a7f7286056ec52b3cd3d432485ca2b9907e0f9/aws-embedded-metrics-3.2.0.tar.gz
```

**Comment by charliermarsh** [comment]
I'm not a fan of omitting the URLs. It makes the installation protocol more complicated (it has to implement the Simple API), and what we're doing here is much more aligned with the standardized `pylock.toml` format from PEP 751 (which also records URLs). Unless I'm misunderstanding, omitting the file URLs doesn't actually solve the problem in this issue anyway, since we still need to record the source, which is what's being modulated?

(As an aside, PyPI's URLs are not exactly opaque identifiers; they're content-addressed locations. The segments comprise the Blake2 hash of the file.)

**Comment by ArcticLampyrid** [comment]
> Unless I'm misunderstanding, omitting the file URLs doesn't actually solve the problem in this issue anyway, since we still need to record the source, which is what's being modulated?

Provide a option to add alternative URLs for sources are easy. But it's hard for sdist/wheel URLs: 

> In the case of the setup we're using, I noticed something which might be useful to call out
> * when the index, is public pypi index i.e. `https://pypi.org/simple`, the package artifacts have `url`s with a prefix looking like `https://files.pythonhosted.org/packages/c1/08/1bb1f7392cd9c3a65c4571a7f7286056ec52b3cd3d432485ca2b9907e0f9/`. For example, for `aws-embedded-metrics`, it's `https://files.pythonhosted.org/packages/c1/08/1bb1f7392cd9c3a65c4571a7f7286056ec52b3cd3d432485ca2b9907e0f9/aws-embedded-metrics-3.2.0.tar.gz`
> * in my case, this custom index, which uses a local proxy on dev/ci machines i.e. `http://<omitted>/pypi/simple`, the package srtifacts `url`s follow a more deterministic url pattern e.g. `http://<omitted>/pypi/simple/<package name>/<package version>/`. The same `aws-embedded-metrics` would be `http://<omitted>/pypi/simple/aws-embedded-metrics/3.2.0/aws-embedded-metrics-3.2.0.tar.gz`.



> As an aside, PyPI's URLs are not exactly opaque identifiers; they're content-addressed locations. The segments comprise the Blake2 hash of the file.

Thanks for information. But I am still curious whether this is a documented detail that can be relied upon.

**Comment by ArcticLampyrid** [comment]
Adding a point, if the purpose of using a mirror is to bypass network censorship issues (which does happen in China due to [GFW](https://en.wikipedia.org/wiki/Great_Firewall)), it is unacceptable to obtain URL details from multiple sources at the same time. Accessing PyPI is likely to result in failure.

**Comment by liblaf** [comment]
Hi everyone, this is a long-standing and important discussion with many thoughtful contributions. As a user who is keenly following this, I wanted to try to synthesize the core problem, the main points of debate, and the proposed solutions so far.

### The Core Problem

The current `uv.lock` file includes absolute URLs for:

- Package registries (`source.registry`)
- Source distributions (`sdist.url`)
- Wheel files (`wheels.url`)

This causes issues when:

1. Developers use different index URLs / proxies locally vs. in CI
2. Organizations have private proxies / mirrors with custom URL patterns
3. Lockfiles contain internal URLs that shouldn't be exposed
4. Environments can't access the originally locked URLs

### The Central Tension: Portability vs. Hermeticity

The discussion has highlighted a key design tension:

- **Portability & Separation of Concerns:** Some feel that a lockfile should define _what_ to install (package, version, hash) but that _where_ it's downloaded from is an installation-time detail. This makes the lockfile portable and keeps environment-specific configuration out of committed project files.
- **Hermeticity & Performance:** Some have emphasized the goal of a hermetic lockfile. Storing exact URLs means `uv sync` doesn't need to perform any index lookups; it can just download the files directly. This is faster, simpler for the installer, and aligns with the direction of the proposed `pylock.toml` standard (PEP 751).

---

Here are the main approaches discussed so far, as I see them:

#### Solution 1: Omit URLs from the lockfile (The "Poetry" / "Separation of Concerns" model)

- **Where it comes from:** This was one of the earliest suggestions by @humanzz and was advocated for by @xbeastx and @ArcticLampyrid. The core idea is that the lockfile should specify _what_ to install (package name, version, hash), while the environment should specify _where_ to get it from. @ArcticLampyrid pointed out that this is how **Poetry** operates.
- **Pros:**
  - **Maximum Portability:** The `uv.lock` file would be completely agnostic to the index/mirror setup and would work seamlessly across any environment.
  - **Simplicity:** Solves all the different use cases (local proxies, different CI URLs, public vs. private URLs) without complex configuration.
- **Cons:**
  - **Not Hermetic:** As @charliermarsh pointed out, this goes against the goal of having a lockfile that can be used for installation without any further network queries to an index API. `uv sync` would need to re-resolve file locations.
  - **Standard Alignment:** This approach seems to diverge from the proposed `pylock.toml` format (PEP 751), which includes URLs.

#### Solution 2: Canonical URL with a Proxy override (The current PR approach)

- **Where it comes from:** This was proposed by @charliermarsh and a draft PR was created by @jtfmumm. The idea is to define a canonical URL for an index in `pyproject.toml`, which gets written to the lockfile. A specific environment can then provide a "proxy" URL via an environment variable (`UV_INDEX_<name>_PROXY_URL`) or flag to use instead.
- **Pros:**
  - **Preserves Hermeticity:** The lockfile still contains valid, canonical URLs.
  - **Clear Intent:** Separates the project's official source from an environment-specific override.
- **Cons:**
  - **The PyPI Mirror Problem:** This is the biggest drawback, as highlighted by @apollo13. This approach only works if the file URLs (for wheels/sdists) are on the same domain as the index URL. It **may not work** for PyPI mirrors (like Nexus, Artifactory, or public mirrors) which rewrite `files.pythonhosted.org` URLs to their own domain. This seems to defeat one of the primary use cases for this feature.

#### Solution 3: Full URL Rewriting / Remapping (The "Git" / "Pixi" model)

- **Where it comes from:** This is a more powerful version of the proxying idea, inspired by `git`'s `url.<base>.insteadOf` (suggested by @ngnpope) and **pixi**'s `[mirrors]` configuration (from issue [#9056](https://github.com/astral-sh/uv/issues/9056)).
- **The Idea:** Allow arbitrary URL prefix remapping in the configuration. This would let a user map _both_ the index URL (`https://pypi.org/simple`) and the file host URL (`https://files.pythonhosted.org`) to their mirror equivalents.
- **Pros:**
  - **Solves the PyPI Mirror Problem:** This is flexible enough to handle the case where file URLs are on a different host from the simple index, making it compatible with most real-world proxy/mirror setups.
  - **Powerful & Flexible:** It can handle a wide variety of scenarios, including the complex internal / external URL mapping described by @ngnpope.
- **Cons:**
  - **Increased Complexity:** This is a more complex feature to design, implement, and for users to configure compared to the other options.

---

I've done my best to capture the nuances of this long discussion, but I may have missed or misinterpreted some points. Please feel free to correct me!

Perhaps we could do an informal poll using emoji reactions on this comment:

- **Solution 1: Omit URLs from the lockfile (The "Poetry" / "Separation of Concerns" model):** React with ❤️
- **Solution 2: Canonical URL with a Proxy override (The current PR approach):** React with 🚀
- **Solution 3: Full URL Rewriting / Remapping (The "Git" / "Pixi" model):** React with 👀
- **None of the above / More discussion needed:** React with 🤔

**Comment by liblaf** [comment]
For anyone who need to use PyPI mirrors and don't want to pollute `uv.lock` with mirror URLs, here's a not-so-elegant workaround. Before each `uv` command, we replace the URLs in `uv.lock` with mirror URLs. After `uv` command is finished, the mirror links are replaced with the default `pypi.org` and `files.pythonhosted.org`.

This workaround only works for mirrors which maps original URLs in a pattern similar to:

```
https://pypi.org/simple
-> https://mirrors.cernet.edu.cn/pypi/web/simple
https://files.pythonhosted.org/packages/1d/56/f5d...8fe/liblaf_grapes-0.6.0.tar.gz
-> https://mirrors.cernet.edu.cn/pypi/web/packages/1d/56/f5d...8fe/liblaf_grapes-0.6.0.tar.gz
```

It works fine with CERNET MirrorZ, and should works with most other mirrors as well. Comments on compatible mirror sites or more elegant and general methods are welcome.

During my test, I found that `uv.lock` may miss a few, only a few `package[*]{sdist,wheels[*]}.{size,upload-time}`. Luckily, this doesn't affect functionality. Even if someone re-lock using origin PyPI index, it won't change `uv.lock`.

The following is a fish shell function that wraps `uv`. Similarly, we can implement alias / functions for other shells, such as bash, zsh, etc.

```fish
function uv --wraps uv
    # similar alias / functions can be implemented in other shells, e.g. bash, zsh
    set --local PYPI_MIRROR "https://mirrors.cernet.edu.cn/pypi/web"
    set --local git_root (git rev-parse --show-toplevel)
    set --local uv_lock "$git_root/uv.lock"
    echo "$uv_lock"
    if test -f $uv_lock
        # `sed` can also be used
        sd --fixed-strings 'https://pypi.org' "$PYPI_MIRROR" "$uv_lock"
        sd --fixed-strings 'https://files.pythonhosted.org' "$PYPI_MIRROR" "$uv_lock"
    end
    UV_DEFAULT_INDEX="$PYPI_MIRROR/simple" command uv "$argv"
    if test -f $uv_lock
        # since mirrorz-302 redirects you to your "best" mirror site, the URLs
        # in `uv.lock` does not have to be the same with `${PYPI_MIRROR}`
        sd 'https://[\.\w]+/pypi/web/simple' 'https://pypi.org/simple' "$uv_lock"
        sd 'https://[\.\w]+/pypi/web/packages' 'https://files.pythonhosted.org/packages' "$uv_lock"
    end
end
```

To use mirrors in [Pixi](https://pixi.sh/)-managed projects (which uses `uv` to handle PyPI packages), [a similar workaround](https://github.com/prefix-dev/pixi/issues/4147) can also be used.

**Comment by waketzheng** [comment]
I replace the registry and url by the following script:
```py
#!/usr/bin/env python
"""Auto change registry of uv.lock to be pypi.org

Usage::
    python3 <me>.py
"""

from __future__ import annotations

import functools
import re
import sys
from pathlib import Path

PYPI = "https://pypi.org/simple"
HOST = "https://files.pythonhosted.org"


def main() -> int | None:
    verbose = "--verbose" in sys.argv
    p = Path("uv.lock")
    if not p.exists():
        p = Path("..", p.name)
        if not p.exists():
            if verbose:
                print(f"{p.name} not found, skip.")
            return None
    text = p.read_text("utf-8")
    registry_pattern = r'(registry = ")(.*?)"'
    replace_registry = functools.partial(re.sub, registry_pattern, rf'\1{PYPI}"')
    registry_urls = {i[1] for i in re.findall(registry_pattern, text)}
    download_pattern = r'(url = ")(https?://.*?)(/packages/.*?\.)(gz|whl)"'
    replace_host = functools.partial(re.sub, download_pattern, rf'\1{HOST}\3\4"')
    download_hosts = {i[1] for i in re.findall(download_pattern, text)}
    if not registry_urls:
        raise ValueError(f"Failed to find pattern {registry_pattern!r} in {p}")
    if len(registry_urls) == 1:
        current_registry = registry_urls.pop()
        if current_registry == PYPI:
            if download_hosts == {HOST}:
                if verbose:
                    print(f"Registry of {p} is {PYPI}, no need to change.")
                return 0
        else:
            text = replace_registry(text)
            if verbose:
                print(current_registry, "-->", PYPI)
    else:
        # TODO: ask each one to confirm replace
        text = replace_registry(text)
        if verbose:
            for current_registry in sorted(registry_urls):
                print(current_registry, "-->", PYPI)
    if len(download_hosts) == 1:
        current_host = download_hosts.pop()
        if current_host != HOST:
            text = replace_host(text)
            if verbose:
                print(current_host, "-->", HOST)
    elif download_hosts:
        # TODO: ask each one to confirm replace
        text = replace_host(text)
        if verbose:
            for current_host in sorted(download_hosts):
                print(current_host, "-->", HOST)
    size = p.write_text(text, encoding="utf-8")
    if verbose:
        print(f"Updated {p} with {size} bytes.")
    return 1


if __name__ == "__main__":
    sys.exit(main())
```

**Comment by dariocurr** [comment]
this is still the reason why we cannot use `uv` in our CI

**Comment by mttbernardini** [comment]
Accessing PyPI from China is either slow or impossible because of the GFW.

The usual workflow here is to use a PyPI mirror like the one provided by [Tsinghua University](https://mirrors.tuna.tsinghua.edu.cn/help/pypi/). However, for my colleagues in UK, using the mirror makes no sense since they can just access PyPI directly, so enforcing it in the `uv.lock` is not practical.

Tools like `pdm` (or `pip`) can be instructed to use an alternative default index via its user-config `~/.config/pdm/config.toml`, and since `pdm.lock` doesn't include wheel urls (by default), I can get the benefit of a `pdm sync` using the Tsinghua mirror instead of PyPI, speeding up my workflow.

I would like to see `uv` also respecting the configuration in `~/.config/uv/uv.toml` for the default index, the very page I linked from Tsingua University suggests this:

```toml
[[index]]
url = "https://mirrors.tuna.tsinghua.edu.cn/pypi/web/simple/"
default = true
```

But since `uv` honours the urls in the `uv.lock`, the config above is ignored (the config above is only respected for `uv tool`/`uvx` I believe), and my UX with `uv sync` becomes unbearable.

I'm for either option in:
1. Don't store the urls in the `uv.lock` and always recompute them based on the relevant `uv.toml` and `pyproject.toml` configs.
2. Store the canonical urls in the `uv.lock`, but add a proxying mechanism so that if the `default` index is re-defined (i.e. via user-config `uv.toml`), then the `uv.lock` urls for the default index are ignored and the proxy index is queried instead.

Either approach has a "hermeticity" issue, but the second one moves it to the proxy scenario, leaving the normal scenario to be efficient/hermetic, assuming that proxying may be less common than using default behaviors.

**Comment by jamesharris-garmin** [comment]
> Perhaps we could do an informal poll using emoji reactions on this comment:
> 
>   * **Solution 1: Omit URLs from the lockfile (The "Poetry" / "Separation of Concerns" model):** React with ❤️
> 
>   * **Solution 2: Canonical URL with a Proxy override (The current PR approach):** React with 🚀
> 
>   * **Solution 3: Full URL Rewriting / Remapping (The "Git" / "Pixi" model):** React with 👀
> 
>   * **None of the above / More discussion needed:** React with 🤔

As much as I personally prefer the git/pixi model or even the current PR, my current issue is that we are currently trying to transition our mirrors from one backend mirror provider (sonatype nexus) to another (artifactory) which use different url schemes for actually hosting the files. This means nothing short of solution 1 would survive this use case.

**Comment by jamesharris-garmin** [comment]
Just to clarify our issue: I work with a private repository that was until recently backed wholy by sonatype nexus. I.e. the domain name on the internal network used a round robin load balancer to route calls to the service to one of 8 different nodes running sonatype nexus. As part of a gradual scaling cutover, we are trying to move the backend server for this to artifactory. So our pool in the load balancer currently is 6 nexus nodes and 2 artifactory nodes.

Since there is no standard format for the direct URLs to the wheels/sdists Sonatype and Jfrog chose different internal schemes to provide their packages. This means that any time we resolve that domain name, we have a 1 in 4 chance of resolving to a server that is hosting the file we want in a completely different URL.

Would a possible solution in the installer be to do the following?

```
1. Attempt to download the package directly from the stored url.
2. If the download fails to complete, ask the registry where the archive is stored and fetch the archive from there?
```

This would potentially slow down the initial sync on a system using a mirror, but the stored hash of the artifacts would verify that the artifacts are correct.

An alternative approach would be to add a flag to to the `uv lock` command that would allow you to refresh the urls while keeping all other attributes of the locked data the same.

It seems like, lacking a standard URL for sourced packages, this level of hermesticity is actively hostile to the goal of supporting widely available mirroring configurations, or even the ability to provide lock files as part of publicly distributable projects.

**Comment by jamesharris-garmin** [comment]
The only work around we have been able to use to handle this is, instead of using `uv sync` to setup the project, we use (in bash)

```bash
uv venv
uv pip install -r <(uv export --format requirements.txt --frozen)
```
This will not be as clean or thurough as a uv_sync because it won't remove dependencies from the environment. but it works well enough for CI environments where the goal is to populate a newly created build environment with the installed requirements.

**Comment by notatallshaw** [comment]
> uv venv
> uv pip install -r <(uv export --format requirements.txt --frozen)
> 
> This will not be as clean or thurough as a uv_sync because it won't remove dependencies from the environment.

Are you aware of `uv pip sync`? https://docs.astral.sh/uv/reference/cli/#uv-pip-sync

**Comment by Wim-De-Clercq** [comment]
I like to add that when using a locally cached wheelhouse directory instead of index urls and 
```
export UV_FIND_LINKS=~/.cache/pip-wheelhouse
``` 
as a variable. You end up with with local paths in the lockfile as well. Similar to the index urls.

```
source = { registry = "../../.cache/pip-wheelhouse" }
```
I think this makes even less sense than urls in the lock file. This requires my colleagues and servers to use the same folder structure, not only for the cache folder. But also it is relative to where the project is located in relation to the cache folder.
A local file path should not end up in a lock file.

**Comment by jamesharris-garmin** [comment]
> > uv venv
> > uv pip install -r <(uv export --format requirements.txt --frozen)
> > This will not be as clean or thurough as a uv_sync because it won't remove dependencies from the environment.
> 
> Are you aware of `uv pip sync`? https://docs.astral.sh/uv/reference/cli/#uv-pip-sync

I was not aware of this and this basically covers my workaround case. Thanks @notatallshaw!

 But again. This needs to be configurable or the hermesticity vs. reproducibility problem will break the ability to share lockfiles and properly mirror pypi with all available public solutions.

**Comment by powercoconola** [comment]
> this is still the reason why we cannot use `uv` in our CI

Agreed, hoping this is solved soon. No one can use my projects I build due to my custom `UV_INDEX` littered in the lockfiles and `pyproject.toml`. I am having to go back and update this.

I initially wanted to set `UV_INDEX` for use with `uv tool install` only, but did not know it also affected `uv lock`, etc. Maybe we can have some separation and have a `UV_TOOL_INDEX`?

**Comment by lorenzosciarra** [comment]
I also have a use case in which urls in lock file prevent me from switching from PDM to UV (actually there are a couple considering that the 'uv sync -U' doesn't yet update the pyproject.toml pinned versions, but this is another story):

In my project we use different AWS accounts for different environments (Test, Staging, Prod), and we publish shared Python libraries to the corresponding CodeArtifact repositories. Finally we use them as dependencies of other modules. This is just fine with PDM because index urls and credentials can be entirely managed in the pdm.toml, which is not pushed to git, while the pdm.lock file is agnostic and doesn't change. It can't be achieved with UV at the moment instead, because syncing without locking will try to install libraries from the wrong environment, and locking will cause the lock file to continuously change.

**Comment by jamesharris-garmin** [comment]
Did we actually have a mechanism to force re-write the urls from the versions in the lock file while preserving the versions we are interested in?

**Comment by jessekv** [comment]
The big reason I switched from requirements.txt to a lockfile is that I want to use the lockfile's hashes to ensure the integrity of the environment, regardless of the provenance. Sometimes, where I am installing, I don't have an easy template-friendly URL transformation (or even a URL to transform to).

**Comment by jamesharris-garmin** [comment]
this is less than ideal because it requires checking in changes but if you need to refresh the urls for packages you can run `uv lock --refresh` with your mirror settings and it will re-write the lock file's url lookups to point to the correct locations.

I am not sure if this ensures hashing.

**Comment by yueyinqiu** [comment]
I'd like to add that this isn't just a lock file issue. Sometimes we need to specify the source for a particular package. For example, specifying a specific version of PyTorch, which is typically done in uv by adding the corresponding source to pyproject.toml.

However, as mentioned earlier, accessing this source might be difficult in some network environments, but we don't want to change the source to a specific mirror (as it might cause inaccessibility on other networks).

Currently, we lack a mechanism to use a specific mirror only locally.

**Comment by axellpadilla** [comment]
Just supporting this, so proxies works correctly, a symple flag on sync that allows resolving again without changing packages versions should work.

**Comment by zsol** [comment]
Thanks for the detailed examples here. We’re evaluating a design intended to preserve one portable lockfile (where the selected versions and hashes remain authoritative) while allowing different machines to access the same canonical index through different proxies.

The current shape is:
- Configure a proxy of a canonical index to a proxy at the project, user, or system level.
- Resolve and download through the proxy. Validate artifacts against the lockfile as usual.
- Continue writing canonical registry and artifact URLs, plus hashes, to uv.lock; proxy URLs would not appear in the lockfile.
- When creating or updating a lock through the proxy, we would also query the canonical index to validate the result selected through the proxy - this is a metadata-only query.

Before committing to this design, I’d like to check it against the deployment setups represented here. If you’re affected, could you share:
1. Does the proxy always serve a strict subset of the artifacts of the canonical index? Or does it serve extra content?
1. Can the canonical index be reached when creating or updating the lock, or must all traffic remain behind the proxy?
1. How often is the lock generated both behind and outside the proxy—for example, by developers locally and Renovate in CI?
1. Must a lock generated outside the proxy always install behind it, or is failure acceptable when the proxy rejects a locked artifact?
1. Would retaining canonical URLs and hashes in uv.lock solve the portability and internal-URL disclosure problem for you?

(We're still iterating on the design; this is validation rather than a shipping commitment.)

**Comment by apollo13** [comment]
> Does the proxy always serve a strict subset of the artifacts of the canonical index? Or does it serve extra content?

We used to have proxies in the past that also injected our own packages. We stopped doing that since it is horrible from a supply chain perspective (better to configure multiple indices and tell the tooling what lives where). In our case the proxies are mirroring pypi 1:1 (on demand)

> Can the canonical index be reached when creating or updating the lock, or must all traffic remain behind the proxy?

On dev machines it can be reached, on CI we'd like to turn it off.

> How often is the lock generated both behind and outside the proxy—for example, by developers locally and Renovate in CI?

exactly, devs locally without proxy and renovate in CI (preferably with proxy).

> Must a lock generated outside the proxy always install behind it, or is failure acceptable when the proxy rejects a locked artifact?

failure is accepted, though there shouldn't be any possibilities of failure if it is a mirror 

> Would retaining canonical URLs and hashes in uv.lock solve the portability and internal-URL disclosure problem for you?

Yes.

**Comment by ziqin** [comment]
> Can the canonical index be reached when creating or updating the lock, or must all traffic remain behind the proxy?

Bad network connection to the canonical index is one of the reasons why many people in Chinese Mainland use mirrored PyPI even on dev machines.

**Comment by yueyinqiu** [comment]
> Does the proxy always serve a strict subset of the artifacts of the canonical index?

Does this imply that the URLs need to be consistent? Not confirmed, but according to https://github.com/prefix-dev/pixi/issues/5486#issuecomment-4321724459, it seems that some might use a completely different URL pattern.

I'm not sure if we should consider this, as this URL pattern can be completely arbitrary. (It seems the only solution is to let proxies to simultaneously provide a mapping from their URLs to the official URLs. But if that's the case, it feels better to go a step further and design a dedicated package source protocol for uv.)

(I've ruled out getting the URL from the canonical index. Network issues are likely the main reason for needing to use a proxy; we shall assume the canonical index is inaccessible.)

**Comment by zsol** [comment]
> Does this imply that the URLs need to be consistent? 

No, the wheels themselves can be hosted anywhere, as long as their filenames (the final part of the URL) and hashes match the canonical index. For example a proxy index that redirects the artifacts to a content-addressed storage should work with this design.

**Comment by nate-england** [comment]
Is it possible to explain why urls need to be stored in the lock? I agree with others that still providing access to the canonical upstream is not ideal. For the users just looking to use a proxy for better reliability, it's fine. For those locked down strictly by security policies, it's a deal breaker. Given that the lockfile contains package names and hashes, it's not clear to me what value the canonical url provides in terms of integrity of the lockfile.

**Comment by axellpadilla** [comment]
Hi @zsol, my answers:

**Does the proxy always serve a strict subset of the artifacts of the canonical index? Or does it serve extra content?**
R.// It could, but we should respect the uv.lock versions for security.

**Can the canonical index be reached when creating or updating the lock, or must all traffic remain behind the proxy?**
R.// When we use the proxy, the proxy can be used as a security trap, blocking unwanted packages/versions, this means, canonical could be blocked.

**How often is the lock generated both behind and outside the proxy—for example, by developers locally and Renovate in CI?**
R.// We really dont need to regenarate it, for that we can just delete it, currently our workaround is this:

`uv lock check && rm uv.lock && uv sync`

(take this as seudo code if something wrong)

**Must a lock generated outside the proxy always install behind it, or is failure acceptable when the proxy rejects a locked artifact?**
R.// Failure is acceptable, because the proxy could be used for security reasons, that said, this is extra work, but sometimes one version could be blocked/yankeed but next minor could work, if this flow exists to use the next minor this will be appreciated.

**Would retaining canonical URLs and hashes in uv.lock solve the portability and internal-URL disclosure problem for yo**
R.// The presence or not of the canonical URLs is not a problem on our case, just uv sync --locked working, like uv sync --locked --regenerate-urls or ---alow-proxy, etc.

**Comment by zsol** [comment]
> Is it possible to explain why urls need to be stored in the lock?

The gist is that storing full URLs in the lock makes:
- the lockfile a self-contained descriptor of the result of the resolution
- it possible to do installation purely from this descriptor in a trivial way
- installation use half the number of network requests

These are generally desirable properties (orthogonal to integrity), so it's highly unlikely URLs stored in lockfiles will go away.

The design I mentioned above would make uv discard all but the filename part of these URLs when a proxy is in effect, so for an installation it would need to go to the (proxy) index and fetch the (proxied) URLs of the required artifacts. To be clear, in these cases, there are no network requests to the origin index.

The only time uv would make network requests to the origin index is when writing a new lockfile, to ensure that the resolved artifacts (resolved via the proxy) are actually available on the origin index too. This preserves the guarantee that the lockfile works without the proxy too.

There's an option to remove this guarantee - part of what I'm trying to understand here is if that option would unblock a substantial amount of use cases (sounds like yes, so far).

**Comment by yueyinqiu** [comment]
I personally don't object to including URLs in the lock file, but since we've brought this up, I'm kind of wondering why these three points are important.

> the lockfile a self-contained descriptor of the result of the resolution
> it possible to do installation purely from this descriptor in a trivial way

As I understand it, the result of the resolution should be uniquely determined by `package name + version`, without needing to include the URL? (As mentioned, it's orthogonal to integrity and will even be ignored when proxying, so it doesn't seem to offer much help? Perhaps there's no reason to provide two different implementations for proxy and canonical sources.)

I'd also like to know if we plan to support proxying for custom sources (e.g., specifying the official PyTorch repository in `pyproject.toml`, but accessing it using a proxy). If so, how will this URL be inferred when generating the lock file using only the proxy?

> installation use half the number of network requests

The majority of time and bandwidth should be spent downloading the package itself. Would recording the URL be of much help?

---

To reiterate, I'm more out of curiosity than against storing URLs (as long as it doesn't cause any problems).

**Comment by jessekv** [comment]
> No, the wheels themselves can be hosted anywhere, as long as their filenames (the final part of the URL) and hashes match the canonical index. For example a proxy index that redirects the artifacts to a content-addressed storage should work with this design.

Would this also include a local directory of pre-downloaded wheels? (somewhat related: https://github.com/astral-sh/uv/issues/3163)

**Comment by zsol** [comment]
> The majority of time and bandwidth should be spent downloading the package itself. Would recording the URL be of much help?

This is not, generally speaking, true. Hitting the index on the download path can take anywhere from a "measurable" to "substantial" portion of the time it takes for a cold `uv sync` operation, depending on your exact setup (number and size of packages, index & file-serving performance).

> I'd also like to know if we plan to support proxying for custom sources (e.g., specifying the official PyTorch repository in `pyproject.toml`, but accessing it using a proxy). If so, how will this URL be inferred when generating the lock file using only the proxy?

Yes, this will be supported. URLs are not inferred by uv. The proxy index URL is in configuration, and the download artifacts are given directly by the proxy index. When writing a new lockfile using the proxy, the URLs within will be from the canonical index (unless the option discussed above is turned on).


> > No, the wheels themselves can be hosted anywhere, as long as their filenames (the final part of the URL) and hashes match the canonical index. For example a proxy index that redirects the artifacts to a content-addressed storage should work with this design.
> 
> Would this also include a local directory of pre-downloaded wheels? (somewhat related: [#3163](https://github.com/astral-sh/uv/issues/3163))

Not initially, although I've explored this option and it _should_ be possible to add later on.

**Comment by apollo13** [comment]
> Yes, this will be supported. URLs are not inferred by uv. The proxy index URL is in configuration, and the download artifacts are given directly by the proxy index.

What does "in configuration" mean here? Or rather: which configuration. We'd like to keep the proxy URL (for the PyPI mirror case)  out of pyproject.toml. Reasoning is having an OSS project that we also run against CI internally where the proxy should be used but the outside world should be able to resolve against PyPI and we don't want to leak our internal infra.

**Comment by zsol** [comment]
> What does "in configuration" mean here? Or rather: which configuration.

It can be in the system-wide, user-wide `uv.toml`, or the project configuration. Cascading works as usual (project overrides user, which overrides system)

**Comment by yueyinqiu** [comment]
> When writing a new lockfile using the proxy, the URLs within will be from the canonical index (unless the option discussed above is turned on).

Hmmm... So when this option is turned on, what will the URL be? (Personally, I don't expect it to be the proxy's URL. I'm ok with a certain special value or just left blank.)

(I failed to find the relevant details; sorry if I have missed something.)

> It can be in the system-wide, user-wide uv.toml, or the project configuration. Cascading works as usual (project overrides user, which overrides system)

Will a local-project-level configuration be required that can override the project-level one? (to allow different proxies to be used for different projects) This might not be necessary for official repository, but could be useful for some custom repositories because they might have the same name in different `pyproject.toml` files?

**Comment by axellpadilla** [comment]
@yueyinqiu a proxy generates its own cache and because that, urls, at least mine, but this is transparent for uv, it justs uses those

**Comment by ArcticLampyrid** [comment]
We should probably have two modes:

* **Alternative Server Mode:**
  * Uses the index from the mirror.
  * Downloads wheels from the mirror.
  * If the download fails, the package installation fails.


* **Cache Acceleration Mode:**
  * Uses the index from the canonical server to determine concepts like "latest version" and to get the correct hashes of the packages.
  * If the package (matching version) exists on the mirror, it downloads the corresponding wheel from the mirror and verifies whether it has been modified by checking its hash against the canonical index.
  * If the package does not exist on the mirror, or if the download from the mirror fails, it falls back to downloading from the canonical server.


These two modes are suitable for different scenarios. "Cache Acceleration" is mainly used for speedup purposes when utilizing an untrusted cache source (without concerning ourselves with issues like IP information leakage).

**Comment by ArcticLampyrid** [comment]
> > When writing a new lockfile using the proxy, the URLs within will be from the canonical index (unless the option discussed above is turned on).
> 
> Hmmm... So when this option is turned on, what will the URL be? (Personally, I don't expect it to be the proxy's URL. I'm ok with a certain special value or just left blank.)

I'm also curious about this. Perhaps PyPI's content-addressed URLs are predictable. But what if we set up a proxy source (M) for a third-party source (A)? It's even possible that both M and A are using their own custom, unpredictable addresses (such as OSS object IDs?).

**Comment by GroenteLepel** [comment]
I'm really curious about the status this feature, since we started using our proxy for all our packages since recently. This means that our .lock file on our open-source packages are useless for us as maintainers, because we cannot install with it. It's a bit hard to get an overview of the track of this issue since I've found quite some open issues and one closed PR:

- #15741 (newest)
- #6349 (original design discussion) 
- #8262 
- #8076 (behavior around lockfile URLs and fallback indexes)

 - https://github.com/astral-sh/uv/pull/11782 (the closed pr)

So am I safe to assume that this _will_ be implemented, it's just not picked up yet? If not, is there anything I provide to further help with this?

Because right now our solution kind of defeats the purpose of a .lock file...
 - create .lock, add it to github
 - let dependabot take care of that .lock file for contributors
 - if we (maintainers) develop, pull, then
 - `git update-index --skip-worktree uv.lock`
 - `uv lock --upgrade`
 - develop
 - regularly _manually_ run `uv lock --upgrade` locally
 - any work on other branches repeat entire proces

**Comment by zsol** [comment]
Yes, I'm working on this. There are lots of nuances here, and I can tell you upfront that it's impossible to solve for _every_ situation raised in this and related issues, but I intend to ship something that can solve the majority of them

**Comment by yaryna1499** [comment]
Hey, guys!
First of all, thank you for building `uv`—it's been a fantastic tool to work with.

We're seeing what seems like a related issue, but from a slightly different angle.

In our environment, developers and automation intentionally use different package proxies with different access policies.

For example:

- Developers use a proxy that blocks packages newer than 10 days.
- Dependabot uses a different proxy without the 10-day quarantine so it can deliver security updates as soon as they become available.

(I know that package proxies often support allow lists and there are ways to work around this, but that's not the point of this question.)

This leads to the following scenario:

1. Dependabot updates `foo` to `1.2.3` and regenerates `uv.lock`.
2. The PR is merged.
3. Before `foo==1.2.3` becomes older than 10 days, a developer runs:

```bash
uv add new-package
```

Although the developer is only adding an unrelated dependency, `uv` appears to re-resolve the lock file using the developer's registry configuration.

The proxy rejects `foo==1.2.3` because it is still within the quarantine window, causing `uv add` to fail.

The issue for us isn't which registry is written into `uv.lock`, but that adding an unrelated dependency requires re-validating already locked packages against the current registry configuration.

Would this be considered the same underlying issue, or would it make sense to open a separate issue specifically about preserving metadata for packages that aren't being modified?

For comparison, this isn't behavior we've encountered with `npm`, where adding an unrelated dependency doesn't require existing locked dependencies to be revalidated against the current registry configuration.

**Comment by zsol** [comment]
@yaryna1499 this is an example of a case that will probably not be solved by the changes I'm working on. It’s less about rewriting index URLs and more about using a lock produced under one package-availability policy under another.

The core issue is that when uv changes the lockfile, it needs to establish that the existing resolution is still valid under the active index configuration. In your example, that configuration intentionally says that `foo==1.2.3` is unavailable. Preserving it anyway would relax the guarantee that the newly written `uv.lock` is valid and installable under the current package policy.

The way I've seen this solved is to configure dependabot and/or uv itself (using [exclude-newer](https://docs.astral.sh/uv/concepts/indexes/#configuring-exclude-newer-for-an-index)) with the same package policy as the proxy.

**Comment by yueyinqiu** [comment]
@yaryna1499 `uv add` seems to do more than just update the lockfile; it actually installs the package onto the device. I'm a bit curious, in this case, the developer's device certainly won't have access to this new version of the package, so what's the expected procedure?

**Comment by rakhimovv** [comment]
Adding a data point for the design validation above, since our deployment is exactly the shape described.

**Setup.** An enterprise artifact-repository PyPI proxy (Artifactory), configured machine-wide through `UV_DEFAULT_INDEX` rather than per project. The projects we consume are third-party open-source repos that ship a `uv.lock` resolved against pypi.org, and whose READMEs tell you to run `uv sync --locked`. That command fails immediately on any of our machines:

```
error: The lockfile at `uv.lock` needs to be updated, but `--locked` was provided.
```

The error does not mention the index, which is what makes it expensive to diagnose the first time — that part is already tracked in #17138.

**The re-lock is not cosmetic, which surprised us.** I expected the regenerated lockfile to differ only in registry URLs. It does not. Measured on one such project (360 packages, `requires-python = ">=3.10, <3.13"`), re-locking with only the index changed:

| | committed lock (pypi.org) | re-locked through the proxy |
| --- | --- | --- |
| package count / versions | 360 | 360, all identical |
| unique `sha256` hashes | 3323 | 1076 |
| macOS / Windows / aarch64 wheel entries | 1375 | 0 |

No package changed name or version, and no hash appears in the regenerated lock that was not already in the committed one — it is strictly a subset. But every non-`linux_x86_64` wheel is gone, so a universal lockfile becomes a single-platform one. Reproduced twice, restoring the original in between.

This is not the proxy serving fewer files: for one spot-checked package, `pypi.org/simple/numpy/` and the proxy's index both list the same 35 files for 1.26.4. I have not worked out *why* the resolution narrows, and I did not want to guess in public — it should reproduce against any PyPI proxy.

The practical consequence is worse than the failed `--locked` check. A contributor who takes the error message's advice and runs `uv lock` produces a lockfile that still resolves and installs correctly for them, and quietly drops platform coverage for everyone else. That is the kind of change that gets committed.

**On the proposed design.** Resolving and downloading through the proxy while writing canonical registry and artifact URLs plus hashes to `uv.lock` would solve our case completely, and it would also remove the failure mode above, since there would no longer be a reason to regenerate the lock at all.

**What we do today,** for anyone arriving here from a search: override `UV_DEFAULT_INDEX` back to `https://pypi.org/simple` for the specific command, rather than in a shell profile, so it cannot leak into a later `uv lock`. That is only viable because our hosts can reach `files.pythonhosted.org` directly — as noted in #19625, uv downloads from the URLs recorded in the lock, so a pypi.org-locked project bypasses the proxy at install time regardless. On a genuinely isolated network there is no override that helps, which is the other half of what the proposed design would fix.
